sdk/core/azure-core/src/main/java/com/azure/core/credential/SimpleTokenCache.java

```diff
@@ -3,53 +3,145 @@
-
-    private final AtomicBoolean wip;
-    private AccessToken cache;
-    private final ReplayProcessor<AccessToken> emitterProcessor = ReplayProcessor.create(1);
-    private final FluxSink<AccessToken> sink = emitterProcessor.sink(OverflowStrategy.BUFFER);
+    private final AtomicReference<MonoProcessor<AccessToken>> wip;
+    private volatile AccessToken cache;
+    private volatile OffsetDateTime nextTokenRefresh = OffsetDateTime.now();
     private final Supplier<Mono<AccessToken>> tokenSupplier;
+    private final Predicate<AccessToken> shouldRefresh;
+    private final Duration refreshRetryTimeout;
+    private final ClientLogger logger = new ClientLogger(SimpleTokenCache.class);
 
     /**
      * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
      *
      * @param tokenSupplier a method to get a new token
      */
     public SimpleTokenCache(Supplier<Mono<AccessToken>> tokenSupplier) {
-        this.wip = new AtomicBoolean(false);
+        this(tokenSupplier, new TokenRefreshOptions());
+    }
+
+    /**
+     * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
+     *
+     * @param tokenSupplier a method to get a new token
+     * @param tokenRefreshOptions the options to configure the token refresh behavior
+     */
+    public SimpleTokenCache(Supplier<Mono<AccessToken>> tokenSupplier, TokenRefreshOptions tokenRefreshOptions) {
+        this.wip = new AtomicReference<>();
         this.tokenSupplier = tokenSupplier;
+        this.shouldRefresh = accessToken -> OffsetDateTime.now().isAfter(accessToken.getExpiresAt()
+            .minus(tokenRefreshOptions.getTokenRefreshOffset()));
+        this.refreshRetryTimeout = tokenRefreshOptions.getTokenRefreshRetryTimeout();
     }
 
     /**
      * Asynchronously get a token from either the cache or replenish the cache with a new token.
      * @return a Publisher that emits an AccessToken
      */
     public Mono<AccessToken> getToken() {
-        if (cache != null && !cache.isExpired()) {
-            return Mono.just(cache);
-        }
-        return Mono.defer(() -> {
-            if (!wip.getAndSet(true)) {
-                return tokenSupplier.get().doOnNext(ac -> cache = ac)
-                    .doOnNext(sink::next)
-                    .doOnError(sink::error)
-                    .doOnTerminate(() -> wip.set(false));
+        try {
+            if (wip.compareAndSet(null, MonoProcessor.create())) {
+                final MonoProcessor<AccessToken> monoProcessor = wip.get();
+                OffsetDateTime now = OffsetDateTime.now();
+                Mono<AccessToken> tokenRefresh;
+                Mono<AccessToken> fallback;
+                if (cache != null && !shouldRefresh.test(cache)) {
+                    // fresh cache & no need to refresh
+                    tokenRefresh = Mono.empty();
+                    fallback = Mono.just(cache);
+                } else if (cache == null || cache.isExpired()) {
```

Comment:
All of this should happen lazily when the `Mono` is subscribed. Wrap this whole `try` block with `Mono.defer()`.

Reply:
Fixed